tachyon/crypto/commitments/fri/two_adic_fri_pcs.h

```diff
@@ -0,0 +1,380 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#ifndef TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_PCS_H_
+#define TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_PCS_H_
+
+#include <algorithm>
+#include <map>
+#include <optional>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/bits.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_proof.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_prover.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_verifier.h"
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_tree.h"
+#include "tachyon/crypto/commitments/mixed_matrix_commitment_scheme.h"
+#include "tachyon/math/finite_fields/extension_field_traits_forward.h"
+#include "tachyon/math/geometry/dimensions.h"
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/math/matrix/matrix_utils.h"
+#include "tachyon/math/polynomials/univariate/evaluations_utils.h"
+#include "tachyon/zk/air/plonky3/base/two_adic_multiplicative_coset.h"
+#include "tachyon/zk/air/plonky3/challenger/challenger.h"
+
+namespace tachyon::crypto {
+
+template <typename ExtField, typename InputMMCS, typename ChallengeMMCS,
+          typename Challenger>
+class TwoAdicFriPCS {
+ public:
+  using F = typename math::ExtensionFieldTraits<ExtField>::BaseField;
+  using Coset = zk::air::plonky3::TwoAdicMultiplicativeCoset<F>;
+  using Commitment = typename InputMMCS::Commitment;
+  using ProverData = typename InputMMCS::ProverData;
+  using Proof = typename InputMMCS::Proof;
+  using InputProof = std::vector<BatchOpening<F, InputMMCS>>;
+  using TwoAdicFriProof =
+      crypto::TwoAdicFriProof<ExtField, ChallengeMMCS, F, InputProof>;
+
+  using Points = std::vector<std::vector<ExtField>>;
+
+  using OpenedValuesForMat = std::vector<std::vector<ExtField>>;
```

Comment:
If it's a matrix why don't you use `math::Matrix`?

Reply:
I followed [Plonky3's implementation ](https://github.com/Plonky3/Plonky3/blob/d4422e047148d3f5480b2e94cb92f54027a03fcd/fri/src/two_adic_pcs.rs#L281). Though`math::Matrix` could be used, I think as a vector it's easier to manipulate and implement. `math::Matrix` may be more efficient, though, so I'll put a note to test this in the future.